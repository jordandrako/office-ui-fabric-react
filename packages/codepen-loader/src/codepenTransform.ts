import * as babylon from 'babylon';
import * as prettier from 'prettier';
import * as fs from 'fs';
import * as path from 'path';

import { Recast, JSCodeShift, IASTPath, IASTSpecifier, IASTNode } from './interfaces';

// recast and jscodeshift have no @types :(
const recast: Recast = require('recast');
const jscodeshift: {
  withParser: (parser: string) => JSCodeShift;
} = require('jscodeshift');

const exampleData = fs.readFileSync(path.resolve(__dirname, '../lib/exampleData.ts')).toString();

const j = jscodeshift.withParser('babylon');

const parse = (source: string) =>
  babylon.parse(source, {
    sourceType: 'module',
    plugins: ['jsx', 'typescript', 'classProperties', 'objectRestSpread']
  });

/**
 * This Transform modifies Fabric website code examples into a format that will allow them to be
 * rendered on Codepen, as part of the "Export to Codepen" feature.
 * There are two types of supported example templates:
 *
 *     [imports]
 *     [named variable export with example code inside]
 *
 * and
 *
 *     [imports]
 *     [named class export with example code inside]
 *
 * Currently, examples which are scattered across multiple files are NOT supported.
 *
 * @param file Example file contents to transform
 * @returns The transformed file contents
 */
export function transform(file: string): string {
  let sourceStr = file;
  // If the exampleData file was imported, append the file contents
  if (sourceStr.indexOf('/utilities/exampleData') !== -1) {
    sourceStr += `\n${exampleData}\n`;
  }

  const source = j(recast.parse(sourceStr, { parser: { parse } }));

  // Make a list of imported identifiers, and remove all imports
  const identifiers: string[] = [];
  source.find(j.ImportDeclaration).forEach((p: IASTPath) => {
    const importPath = p.node.source.value;
    // Ignore identifiers from the React import (which will be a global) and from exampleData
    // (since that whole file is appended to the example if needed)
    if (importPath !== 'react' && !/(exampleData|\.scss|\.css)$/.test(importPath)) {
      p.node.specifiers.forEach((spec: IASTSpecifier) => {
        identifiers.push(spec.local.loc.identifierName);
      });
    }

    // Remove the import
    p.prune();
  });

  let exampleName;
  // remove exports and replace with variable or class declarations, whichever the original example used
  source
    .find(
      j.ExportNamedDeclaration,
      (node: IASTNode) =>
        node.declaration.type === 'VariableDeclaration' ||
        node.declaration.type === 'FunctionDeclaration'
    )
    .replaceWith((p: IASTPath) => {
      if (p.node.declaration.type === 'VariableDeclaration') {
        exampleName = p.node.declaration.declarations[0].id.name;
      }
      return p.node.declaration;
    });

  source
    .find(
      j.ExportNamedDeclaration,
      (node: IASTNode) =>
        node.declaration.type === 'ClassDeclaration' ||
        node.declaration.type === 'TSInterfaceDeclaration'
    )
    .replaceWith((p: IASTPath) => {
      if (p.node.declaration.type === 'ClassDeclaration') {
        exampleName = p.node.declaration.id.name;
      }
      return p.node.declaration;
    });

  // Build the list of imports from window.Fabric
  let attachedWindowString = 'const {';
  if (identifiers.length > 0) {
    attachedWindowString += identifiers.join(',') + ',';
  }
  attachedWindowString += 'Fabric} = window.Fabric;\n';

  // add imports and React render footer
  const sourceWithFooter = [
    attachedWindowString,
    source.toSource(),
    `ReactDOM.render(<${exampleName}/>, document.getElementById("content"));`
  ].join('\n');

  return prettier.format(sourceWithFooter, {
    parser: 'typescript',
    printWidth: 120,
    tabWidth: 2,
    singleQuote: true
  });
}
