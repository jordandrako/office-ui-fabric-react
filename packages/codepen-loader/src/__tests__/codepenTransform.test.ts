import { transform } from '../codepenTransform';
import * as fs from 'fs';
import * as path from 'path';

describe('exampleData', () => {
  it('contains all expected interfaces', () => {
    // This test is protection against interfaces/functions being split out from
    // office-ui-fabric-react/src/utilities/exampleData.ts without the loader also being updated.
    const filename = path.resolve(__dirname, '../../lib/exampleData.ts');
    const exampleData = fs.readFileSync(filename).toString();
    expect(exampleData).toContain('export interface IExampleItem');
    expect(exampleData).toContain('export function createListItems');
    expect(exampleData).toContain('export function createGroups');
    expect(exampleData).toContain('export function lorem');
    expect(exampleData).toContain('export function isGroupable');
  });
});

describe('codepen transform', () => {
  function testFile(file: string): void {
    const filename = path.resolve(__dirname, './examples/' + file);
    const fileContents = fs.readFileSync(filename).toString();
    const transformResult = transform(fileContents);
    expect(transformResult).toMatchSnapshot();
  }

  it('handles examples with function components', () => {
    testFile('function.txt');
  });

  it('handles examples with class components', () => {
    testFile('class.txt');
  });

  it('handles examples importing exampleData', () => {
    testFile('exampleData.txt');
  });
});
