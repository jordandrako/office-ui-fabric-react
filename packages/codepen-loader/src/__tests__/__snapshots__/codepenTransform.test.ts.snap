// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codepen transform handles examples importing exampleData 1`] = `
"const { IGroup, Fabric } = window.Fabric;

const FakeExampleDataExample = () => {
  return <div>hi</div>;
};

const LOREM_IPSUM = (
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut ' +
  'labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut ' +
  'aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore ' +
  'eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt ' +
  'mollit anim id est laborum'
).split(' ');

const DATA = {
  color: ['red', 'blue', 'green', 'yellow'],
  shape: ['circle', 'square', 'triangle'],
  location: ['Seattle', 'New York', 'Chicago', 'Los Angeles', 'Portland']
};

interface IExampleItem {
  thumbnail: string;
  key: string;
  name: string;
  description: string;
  color: string;
  shape: string;
  location: string;
  width: number;
  height: number;
}

function createListItems(count: number, startIndex: number = 0): IExampleItem[] {
  return Array.apply(null, Array(count)).map((item: number, index: number) => {
    const size = 150 + Math.round(Math.random() * 100);

    return {
      thumbnail: \`//placehold.it/\${size}x\${size}\`,
      key: 'item-' + (index + startIndex) + ' ' + lorem(4),
      name: lorem(5),
      description: lorem(10 + Math.round(Math.random() * 50)),
      color: _randWord(DATA.color),
      shape: _randWord(DATA.shape),
      location: _randWord(DATA.location),
      width: size,
      height: size
    };
  });
}

function createGroups(
  groupCount: number,
  groupDepth: number,
  startIndex: number,
  itemsPerGroup: number,
  level: number = 0,
  key: string = '',
  isCollapsed?: boolean
): IGroup[] {
  if (key !== '') {
    key = key + '-';
  }
  const count = Math.pow(itemsPerGroup, groupDepth);
  return Array.apply(null, Array(groupCount)).map((value: number, index: number) => {
    return {
      count: count,
      key: 'group' + key + index,
      name: 'group ' + key + index,
      startIndex: index * count + startIndex,
      level: level,
      isCollapsed: isCollapsed,
      children:
        groupDepth > 1
          ? createGroups(groupCount, groupDepth - 1, index * count + startIndex, itemsPerGroup, level + 1, key + index)
          : []
    };
  });
}

function lorem(wordCount: number): string {
  return Array.apply(null, Array(wordCount))
    .map((item: number, idx: number) => {
      return LOREM_IPSUM[idx % LOREM_IPSUM.length];
    })
    .join(' ');
}

function isGroupable(key: string): boolean {
  return key === 'color' || key === 'shape' || key === 'location';
}

function _randWord(array: string[]): string {
  const index = Math.floor(Math.random() * array.length);
  return array[index];
}

ReactDOM.render(<FakeExampleDataExample />, document.getElementById('content'));
"
`;

exports[`codepen transform handles examples with class components 1`] = `
"const { SpinButton, Fabric } = window.Fabric;

class SpinButtonBasicExample extends React.Component<any, any> {
  public render(): JSX.Element {
    return (
      <div style={{ width: '400px' }}>
        <SpinButton
          defaultValue=\\"0\\"
          label={'Basic SpinButton:'}
          min={0}
          max={100}
          step={1}
          iconProps={{ iconName: 'IncreaseIndentLegacy' }}
          incrementButtonAriaLabel={'Increase value by 1'}
          decrementButtonAriaLabel={'Decrease value by 1'}
        />
      </div>
    );
  }
}

ReactDOM.render(<SpinButtonBasicExample />, document.getElementById('content'));
"
`;

exports[`codepen transform handles examples with function components 1`] = `
"const { Label, Fabric } = window.Fabric;

const LabelBasicExample = () => {
  return (
    <div>
      <Label>I'm a Label</Label>
    </div>
  );
};

ReactDOM.render(<LabelBasicExample />, document.getElementById('content'));
"
`;
